### PR TITLE
Added Python and Numpy implementation of getdminfo

### PR DESCRIPTION
**Pull request by astrofrog**
This is still a work in progress, but is close to done. Remaining to-dos:

* [x] Switch the CASA dask reader to use this new function
* [x] Fix test for failing cases - there are some corner cases where the bucket size isn't being read in/computed correctly
* [x] Check that there are no other failures in CI

@keflavich - this brings us one step closer to being able to read in CASA datasets without CASA - however before that can happen I would need to add Python implementations of:

```
    casa_cs = ia.coordsys()
    unit = ia.brightnessunit()
    beam_ = ia.restoringbeam()
```

Given that it wasn't too hard to do ``getdminfo``, it might not be too hard to do, though I can see the coordinate system being a little bit trickier potentially.

In any case, with this PR, ``casa_image_dask_reader`` is now CASA-free - it's mainly the meta-data that still needs CASA.

### CODE REVIEW

**Comment by keflavich** [comment]
I hit something right away:
```
In [3]: cube = SpectralCube.read('G012.80_B6_spw3_12M_spw3.image')
Traceback (most recent call last):
  File "<ipython-input-3-c82499cfe334>", line 1, in <module>
    cube = SpectralCube.read('G012.80_B6_spw3_12M_spw3.image')
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/core.py", line 120, in __call__
    return registry.read(BaseSpectralCube, StringWrapper(filename), *args, **kwargs)
  File "/orange/adamginsburg/miniconda3/envs/casa6_py36/lib/python3.6/site-packages/astropy/io/registry.py", line 523, in read
    data = reader(*args, **kwargs)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_image.py", line 107, in load_casa_image
    data = casa_image_dask_reader(filename)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_image.py", line 354, in casa_image_dask_reader
    dminfo = getdminfo(str(imagename))
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_dminfo.py", line 197, in getdminfo
    return read_tiled_cell_st_man(f)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_dminfo.py", line 29, in wrapper
    result = func(f)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_dminfo.py", line 178, in read_tiled_cell_st_man
    st_man = read_tiled_st_man(f)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_dminfo.py", line 29, in wrapper
    result = func(f)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_dminfo.py", line 148, in read_tiled_st_man
    read_record(f)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_dminfo.py", line 29, in wrapper
    result = func(f)
  File "/orange/adamginsburg/repos/spectral-cube/spectral_cube/io/casa_dminfo.py", line 64, in read_record
    raise NotImplementedError('Support for {0} version {1} not implemented'.format(stype, sversion))
NotImplementedError: Support for b'\x00\x00\x00\x06Record\x00\x00\x00\x01\x00\x00\x00\x1a\x00\x00\x00\nRecordDesc\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00' version 67108864 not implemented

```

**Comment by astrofrog** [comment]
@keflavich - you hit an issue which relates to one of the values in the ``table.f0`` file being 64-bit instead of 32-bit for large arrays. This should be fixed now, but let me know if not!

@e-koch - would you be able to take this for a spin? Basically just trying to parse various CASA images with ``SpectralCube.read`` and letting me know if you run into any issues.

**Comment by e-koch** [comment]
@astrofrog -- Some cubes are working great; others are failing with unexpected file size. I'm unsure if it's related to this PR, though.

The cube that reads in fine is a direct output from `tclean` in CASA 5.6. The cubes that are failing are from `clean` in CASA 4.7 or are the output from a custom linear mosaicing script ran in CASA 6. Here's the traceback for the latter:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-1dc2131e2b83> in <module>
----> 1 cube = SpectralCube.read("M33_ACA_12CO21_1p3kms_excludeasymmbeams_roundbeam.image", format='casa')

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/io/core.py in __call__(self, filename, *args, **kwargs)
    116         kwargs['target_cls'] = BaseSpectralCube
    117         try:
--> 118             return registry.read(BaseSpectralCube, filename, *args, **kwargs)
    119         except IsADirectoryError:  # See note above StringWrapper
    120             return registry.read(BaseSpectralCube, StringWrapper(filename), *args, **kwargs)

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/astropy/io/registry.py in read(cls, format, *args, **kwargs)
    521 
    522         reader = get_reader(format, cls)
--> 523         data = reader(*args, **kwargs)
    524 
    525         if not isinstance(data, cls):

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/io/casa_image.py in load_casa_image(filename, skipdata, skipvalid, skipcs, target_cls, **kwargs)
    108     # read in the data
    109     if not skipdata:
--> 110         data = casa_image_dask_reader(filename)
    111 
    112     # CASA stores validity of data as a mask

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/io/casa_image.py in casa_image_dask_reader(imagename, memmap, mask)
    382         else:
    383             raise ValueError("Unexpected file size for data, found {0} but "
--> 384                              "expected {1} or {2}".format(filesize, totalsize * 4, totalsize * 8))
    385 
    386     # the ratio between these tells you how many chunks must be combined

ValueError: Unexpected file size for data, found 1067404800 but expected 1058932800 or 2117865600
```

`getdminfo` is working in all the cases I looked at:

From `tb.getdminfo()`:

```
{'*1': {'COLUMNS': array(['map'], dtype='<U16'),
  'NAME': 'map',
  'SEQNR': 0,
  'SPEC': {'DEFAULTTILESHAPE': array([64, 77,  6], dtype=int32),
   'HYPERCUBES': {'*1': {'BucketSize': 118272,
     'CellShape': array([1208, 1461,  150], dtype=int32),
     'CubeShape': array([1208, 1461,  150], dtype=int32),
     'ID': {},
     'TileShape': array([64, 77,  6], dtype=int32)}},
   'MAXIMUMCACHESIZE': 0,
   'MaxCacheSize': 0,
   'SEQNR': 0},
  'TYPE': 'TiledCellStMan'}}

```

From `getdminfo()`:

```
{'*1': {'SPEC': {'SEQNR': 0,
   'MAXIMUMCACHESIZE': 0,
   'MaxCacheSize': 0,
   'HYPERCUBES': {'*1': {'CubeShape': array([1208, 1461,  150], dtype=int32),
     'CellShape': array([1208, 1461,  150], dtype=int32),
     'TileShape': array([64, 77,  6], dtype=int32),
     'ID': {},
     'BucketSize': 118272}},
   'DEFAULTTILESHAPE': array([64, 77,  6], dtype=int32)},
  'SEQNR': 0,
  'COLUMNS': array(['map'], dtype='<U16'),
  'NAME': 'map',
  'TYPE': 'TiledCellStMan'}}

```

`pformat(test) == pformat(ref)` gives `True`

**Comment by astrofrog** [comment]
@e-koch - thanks for testing this! The failures you are seeing are likely due to a bug in the implementation of #607 - I can fix this here. Is the getdminfo output you showed above for the case where the file size error happens?

**Comment by e-koch** [comment]
@astrofrog -- Yup! Those are the getdminfo outputs for the failure case.

**Comment by astrofrog** [comment]
@e-koch - could you try again?

**Comment by e-koch** [comment]
@astrofrog - That fixed it!

**Comment by astrofrog** [comment]
I've now pushed some additional changes which replace the following CASA calls:

```python
casa_cs = ia.coordsys()
unit = ia.brightnessunit()
beam_ = ia.restoringbeam()
```

with our own parsing of ``table.dat``. In fact, the parsing machinery is almost (but not quite) general enough to be able to read any CASA table, not just tables in image files (making it general enough is probably another day of work). At this point, the only remaining code that uses CASA to read data is the ``wcs_casa2astropy`` function which translates the CASA coordinates into FITS-WCS. Replacing this should be doable, though at this point maybe worth leaving for a follow-up pull request. I would also estimate about a day of work for this - this might be worth it since after that CASA won't be needed to load in CASA data, which means that it will be possible for example to load it on recent Python versions, or even on Windows.

I've also done a bit of tidying up of the CASA-related code, splitting the dask and WCS utilities into their own files.

@keflavich @e-koch - It would be great if you could try out the latest PR on a variety of ``.image`` files to see if you run into any issues. One thing to note is that since some of the structures are versioned, the reader may fail on e.g. quite old files if we don't implement that specific version, but at least the error message should be explicit about that. We can easily add support for older versions if needed, I just don't have any example files for that.

**Comment by keflavich** [comment]
@astrofrog - if reading tables is truly that straightforward, it's worth doing - Ryan has expressed interest in having that too!  I'll explore this as I find time.

**Comment by astrofrog** [comment]
I've opened an issue about generalizing the table reader: https://github.com/radio-astro-tools/spectral-cube/issues/611

**Comment by e-koch** [comment]
I'm encountering a `NotImplementedError` for all casa image cubes I've tried:

```
In [6]: cube = SpectralCube.read("Brick2Tile2_12CO21_2kms.image")                                                                                                            
---------------------------------------------------------------------------
IsADirectoryError                         Traceback (most recent call last)
~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/core.py in __call__(self, filename, *args, **kwargs)
    117         try:
--> 118             return registry.read(BaseSpectralCube, filename, *args, **kwargs)
    119         except IsADirectoryError:  # See note above StringWrapper

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/astropy/io/registry.py in read(cls, format, *args, **kwargs)
    507                         ctx = get_readable_fileobj(args[0], encoding='binary')
--> 508                         fileobj = ctx.__enter__()
    509                     except OSError:

~/anaconda3/envs/casa_py36/lib/python3.6/contextlib.py in __enter__(self)
     80         try:
---> 81             return next(self.gen)
     82         except StopIteration:

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/astropy/utils/data.py in get_readable_fileobj(name_or_obj, encoding, cache, show_progress, remote_timeout)
    192                 timeout=remote_timeout)
--> 193         fileobj = io.FileIO(name_or_obj, 'r')
    194         if is_url and not cache:

IsADirectoryError: [Errno 21] Is a directory: 'Brick2Tile2_12CO21_2kms.image'

During handling of the above exception, another exception occurred:

NotImplementedError                       Traceback (most recent call last)
<ipython-input-6-df3bed4c990c> in <module>
----> 1 cube = SpectralCube.read("Brick2Tile2_12CO21_2kms.image")

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/core.py in __call__(self, filename, *args, **kwargs)
    118             return registry.read(BaseSpectralCube, filename, *args, **kwargs)
    119         except IsADirectoryError:  # See note above StringWrapper
--> 120             return registry.read(BaseSpectralCube, StringWrapper(filename), *args, **kwargs)
    121 
    122 

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/astropy/io/registry.py in read(cls, format, *args, **kwargs)
    521 
    522         reader = get_reader(format, cls)
--> 523         data = reader(*args, **kwargs)
    524 
    525         if not isinstance(data, cls):

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_image.py in load_casa_image(filename, skipdata, skipvalid, skipcs, target_cls, **kwargs)
     69     # read in coordinate system object
     70 
---> 71     desc = getdesc(filename)
     72 
     73     casa_cs = desc['_keywords_']['coords']

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in getdesc(filename)
    445             raise ValueError('Incorrect magic code: {0}'.format(magic))
    446 
--> 447         return read_table(f, filename)

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in wrapper(f, *args)
     25             return
     26         b = BytesIO(f.read(nbytes - 4))
---> 27         result = func(b, *args)
     28         end = f.tell()
     29         if end - start != nbytes:

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in read_table(f, image_path)
    237     name = read_string(f)  # noqa
    238 
--> 239     table_desc = read_table_desc(f, nrow, image_path)
    240 
    241     return table_desc

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in wrapper(f, *args)
     25             return
     26         b = BytesIO(f.read(nbytes - 4))
---> 27         result = func(b, *args)
     28         end = f.tell()
     29         if end - start != nbytes:

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in read_table_desc(f, nrow, image_path)
    289     desc = {}
    290 
--> 291     desc['_keywords_'] = read_table_record(f, image_path)
    292     desc['_define_hypercolumn_'] = {}
    293 

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in wrapper(f, *args)
     25             return
     26         b = BytesIO(f.read(nbytes - 4))
---> 27         result = func(b, *args)
     28         end = f.tell()
     29         if end - start != nbytes:

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in read_table_record(f, image_path)
    218             records[name] = read_array(f, 'string')
    219         elif rectype == 'record':
--> 220             records[name] = read_table_record(f, image_path)
    221         else:
    222             raise NotImplementedError("Support for type {0} in TableRecord not implemented".format(rectype))

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in wrapper(f, *args)
     25             return
     26         b = BytesIO(f.read(nbytes - 4))
---> 27         result = func(b, *args)
     28         end = f.tell()
     29         if end - start != nbytes:

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in read_table_record(f, image_path)
    218             records[name] = read_array(f, 'string')
    219         elif rectype == 'record':
--> 220             records[name] = read_table_record(f, image_path)
    221         else:
    222             raise NotImplementedError("Support for type {0} in TableRecord not implemented".format(rectype))

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in wrapper(f, *args)
     25             return
     26         b = BytesIO(f.read(nbytes - 4))
---> 27         result = func(b, *args)
     28         end = f.tell()
     29         if end - start != nbytes:

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in read_table_record(f, image_path)
    218             records[name] = read_array(f, 'string')
    219         elif rectype == 'record':
--> 220             records[name] = read_table_record(f, image_path)
    221         else:
    222             raise NotImplementedError("Support for type {0} in TableRecord not implemented".format(rectype))

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in wrapper(f, *args)
     25             return
     26         b = BytesIO(f.read(nbytes - 4))
---> 27         result = func(b, *args)
     28         end = f.tell()
     29         if end - start != nbytes:

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in read_table_record(f, image_path)
    189         raise NotImplementedError('Support for {0} version {1} not implemented'.format(stype, sversion))
    190 
--> 191     records = read_record_desc(f)
    192 
    193     unknown = read_int32(f)  # noqa

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in wrapper(f, *args)
     25             return
     26         b = BytesIO(f.read(nbytes - 4))
---> 27         result = func(b, *args)
     28         end = f.tell()
     29         if end - start != nbytes:

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/casa_low_level_io.py in read_record_desc(f)
    176             read_int32(f)
    177         else:
--> 178             raise NotImplementedError("Support for type {0} in RecordDesc not implemented".format(rectype))
    179 
    180     return records

NotImplementedError: Support for type arrayfloat in RecordDesc not implemented

```

The `IsDirectoryError` goes away if `format='casa'` is given (or `casa_image`), though the other cases aren't throwing this error before the `NotImplementedError` here:

```
~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/spectral_cube-0.0.0-py3.6.egg/spectral_cube/io/core.py in __call__(self, filename, *args, **kwargs)
    116         kwargs['target_cls'] = BaseSpectralCube
    117         try:
--> 118             return registry.read(BaseSpectralCube, filename, *args, **kwargs)
    119         except IsADirectoryError:  # See note above StringWrapper
    120             return registry.read(BaseSpectralCube, StringWrapper(filename), *args, **kwargs)
 ```

**Comment by astrofrog** [comment]
@e-koch - thanks for testing this! Adding support for arrayfloat should be an easy fix. Would you be able to send one of the problematic ``table.dat`` files to me by email?

**Comment by astrofrog** [comment]
By the way it seems that the reason I didn't catch this is because of broken behavior in ``putkeywords`` which is even documented here:

https://casa.nrao.edu/docs/CasaRef/table.putkeywords.html#x954-9700003.2.1

I'll investigate workarounds.

File: spectral_cube/io/casa_low_level_io.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+    # Not sure what the following value is
+    nrec = read_int32(f)
+
+    records = OrderedDict()
+
+    for i in range(nrec):
+        name = read_string(f)
+        rectype = TYPES[read_int32(f)]
+        records[name] = {'type': rectype}
+        if rectype == 'bool':
```
Comment (keflavich) [inline spectral_cube/io/casa_low_level_io.py]: not sure if it's worth it, but some of this could be shrunk down by doing 
`if rectype in ('bool', 'int, 'uint'): records[name]['value'] = read_int32(f)` and similar

File: spectral_cube/io/tests/test_casa_dask.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    CASA_INSTALLED = True
+except ImportError:
+    try:
+        from taskinit import ia as image
+        CASA_INSTALLED = True
+    except ImportError:
+        CASA_INSTALLED = False
+
+# NOTE: the (127, 337, 109) example is to make sure that things work correctly
+# when the shape isn't a mulitple of the chunk size along any
```
Comment (keflavich) [inline spectral_cube/io/tests/test_casa_dask.py]:
```suggestion
# when the shape isn't a multiple of the chunk size along any
```

File: spectral_cube/io/casa_dask.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        # The offset is in the final bit array - but fromfile needs to operate
+        # by reading in uint8, so we need to make sure we align what we read
+        # in to the bytes.
+        start = floor(self._offset / 8)
+        end = ceil((self._offset + self._count) / 8)
+        array_uint8 = np.fromfile(self._filename, dtype=np.uint8,
+                                  offset=start, count=end - start)
+        array_bits = np.unpackbits(array_uint8, bitorder='little')
+        chunk = array_bits[self._offset - start * 8:self._offset + self._count - start * 8]
+        return chunk.astype(np.bool_).reshape(self.shape[::-1], order='F').T[item]
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply spectral_cube/io/casa_dask.py]: Good point!

[the remark this answers is not in the document]
Comment (astrofrog) [reply spectral_cube/io/casa_dask.py]: Fixed

File: spectral_cube/io/casa_low_level_io.py (hunk 1 above)
Comment (astrofrog) [reply spectral_cube/io/casa_low_level_io.py]: Done

**Comment by astrofrog** [comment]
@e-koch - could you try again? I think the floatarray issue should be fixed now.

**Comment by e-koch** [comment]
@astrofrog - Loading and basic operations are working! Including on images from older CASA versions (at least v4.7; I don't have anything older on-hand right now).

I expanded the cubes I was testing to all the cube outputs produced by tclean and found a couple minor issues:

1. Some cubes don't have a unit defined, so reading the unit should be optional:
```
KeyError                                  Traceback (most recent call last)
<ipython-input-23-6eccf8d8e90f> in <module>
----> 1 cube = SpectralCube.read("Brick3Tile5_12CO21_1p3kms.mask", format='casa')

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/io/core.py in __call__(self, filename, *args, **kwargs)
    116         kwargs['target_cls'] = BaseSpectralCube
    117         try:
--> 118             return registry.read(BaseSpectralCube, filename, *args, **kwargs)
    119         except IsADirectoryError:  # See note above StringWrapper
    120             return registry.read(BaseSpectralCube, StringWrapper(filename), *args, **kwargs)

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/astropy/io/registry.py in read(cls, format, *args, **kwargs)
    521 
    522         reader = get_reader(format, cls)
--> 523         data = reader(*args, **kwargs)
    524 
    525         if not isinstance(data, cls):

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/io/casa_image.py in load_casa_image(filename, skipdata, skipvalid, skipcs, target_cls, **kwargs)
     73     casa_cs = desc['_keywords_']['coords']
     74 
---> 75     unit = desc['_keywords_']['units']
     76 
     77     if 'perplanebeams' in desc['_keywords_']['imageinfo']:

KeyError: 'units'

```

2. Beam info can also be missing:
```
KeyError                                  Traceback (most recent call last)
<ipython-input-31-d1f60714e353> in <module>
----> 1 cube = SpectralCube.read("Brick3Tile5_12CO21_1p3kms.model", format='casa')

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/io/core.py in __call__(self, filename, *args, **kwargs)
    116         kwargs['target_cls'] = BaseSpectralCube
    117         try:
--> 118             return registry.read(BaseSpectralCube, filename, *args, **kwargs)
    119         except IsADirectoryError:  # See note above StringWrapper
    120             return registry.read(BaseSpectralCube, StringWrapper(filename), *args, **kwargs)

~/anaconda3/envs/casa_py36/lib/python3.6/site-packages/astropy/io/registry.py in read(cls, format, *args, **kwargs)
    521 
    522         reader = get_reader(format, cls)
--> 523         data = reader(*args, **kwargs)
    524 
    525         if not isinstance(data, cls):

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/io/casa_image.py in load_casa_image(filename, skipdata, skipvalid, skipcs, target_cls, **kwargs)
     81         beam_['beams'] = {key: {'*0': value} for key, value in list(beam_['beams'].items())}
     82     else:
---> 83         beam_ = desc['_keywords_']['imageinfo']['restoringbeam']
     84 
     85     wcs = wcs_casa2astropy(data.ndim, casa_cs)

KeyError: 'restoringbeam'

```

**Comment by astrofrog** [comment]
@e-koch - thanks for the further testing! Sorry to bug you, but can you try again now?

**Comment by e-koch** [comment]
@astrofrog - No problem! That fixed it. I now only encounter the error from #608, which occurs after the image read.

I'll finish reviewing now.

**Comment by e-koch** [comment]
#608 could be addressed here, given there are a few changes already. I won't be able to finish adding the maskless test case for a couple of days.

**Comment by astrofrog** [comment]
@e-koch - I've now included your commit from #608 along with a regression test

**Comment by e-koch** [comment]
@astrofrog - thanks! I'll close #608

**Comment by astrofrog** [comment]
@e-koch - thanks for reviewing!
